### DISCUSSION TITLE
Completion behavior with `nvim-cmp` (Neovim 0.6.1)

### DISCUSSION THREAD

**Discussion by folliehiyuki** [ansible/ansible-language-server#230] [Q&A] [answered]
Hi! I'm opening this discussion following the issue https://github.com/hrsh7th/nvim-cmp/issues/615 .

Multiple users of `nvim-cmp`, including me, have reported that when using `nvim-cmp` (has insertReplaceSupport capability) with `ansible-language-server` and `nvim-lspconfig`, the completion text gets removed when scrolling through the completion list. The behavior can be seen below:

https://user-images.githubusercontent.com/67634026/154010580-e482e9e6-53a9-4a03-bbd6-03d00645278c.mp4

Here is an example of how we want it to work (below is with `sumneko_lua`) - that is, the text under cursor gets replaced by the item being hovered in the list:

https://user-images.githubusercontent.com/67634026/154009795-6a45eb29-0b71-437c-a256-aa5a7ef68f6a.mp4

As explained by `nvim-cmp`'s developer in the linked issue, the problem is `ansible-language-server` sends an empty `textEdit.newText` until the item gets resolved. As such, for the 1st time the item is hovered, `nvim-cmp` will treat it as removing text and empty the cursor text.

So, is this behavior intended by `ansible-language-server` or a bug? If it is a latter one, where should it be fixed (on `nvim-cmp` side or `ansible-language-server` side and how)?

**Comment by webknjaz** [comment]
Thanks for the question! I don't know for sure so I'm inviting @yaegassy (the author of https://github.com/yaegassy/coc-ansible) to take a look.

**Comment by yaegassy** [comment]
I don't know because I am not using ansible-language-server with nvim-lspconfig.

By the way, there is no problem with coc-ansible.

https://user-images.githubusercontent.com/188642/154022494-225d0775-fe49-41e5-a381-7ef554dc8cf7.mp4

---

**Edit**

> As a side note, I checked my DEMO with `"ansible.ansible.useFullyQualifiedCollectionNames": false` settings.

**Comment by folliehiyuki** [comment]
Sorry @hrsh7th for the ping! Can you look into this discussion and clarify how you think `nvim-cmp` should work (since I'm just an user and pretty ignorant)?

`coc.nvim` works fine here, so should we reopen the issue in nvim-cmp?

**Reply by yaegassy** [reply]
I don't think this is a bug in the Language Server. Also, it is a specific behavior of Vim and Vim plugin that text is inserted into the document when the completion list is selected. (VSCode behaves differently)

Also, in your nvim built-in lsp with ansible-language-server demo, it looks like a problem on the `nvim built-in lsp` or `nvim-cmp` side, because when I scroll up, it inserts the already expanded text.

In coc.nvim, we simply display the `label`. I think nvim-cmp is looking at `newText`, which is also the correct behavior.

**Comment by hrsh7th** [comment, marked as answer]
OK. I figure out the reason.
The ansiblels returns `{ newText: "", range: ... }` as first response and return `{ newText: "containers.podman.podman_container:", range: ... }` as second response for `completionItem/resolve`.

I don't think this is valid for the spec, but VSCode/coc.nvim seems to support it, so I'll consider fixing it.

**Reply by hrsh7th** [reply]
Hmm ... I don't think VS Code and coc.nvim work the same way.
I'll check the VSCode's behavior.

I think this discussion can be closed. I'll continue this discussion at the nvim-cmp's issue.

**Reply by hrsh7th** [reply]
Sorry! It's my misake. The VSCode also shows the items and coc.nvim works properly.

**Reply by hrsh7th** [reply]
I've fixed in nvim-cmp.
https://github.com/hrsh7th/nvim-cmp/issues/615

**Comment by folliehiyuki** [comment]
Thanks everyone for participating! Closing since it is fixed.
